### PR TITLE
Blue green deployment script

### PR DESCRIPTION
**Pull request by thebowenfeng**
Zero downtime deployment script using blue-green deployment strategy. All nodes have both "blue" instances (old) and "green" instances (new) running simultaneously during deployment, and after a set amount of time blue instances are killed and removed.

Caddy is used as a reverse proxy. Script uses SSH to remotely run commands on each node.

Sample usage of script at top of file. Script must be executed in the same instance where Caddy is being hosted.

### CODE REVIEW

**Comment by thebowenfeng** [comment]
Not _quite_ ready yet, just need to figure out how to tell Caddy to stop routing traffic to old nodes (but obviously still complete any existing requests, hence the 30s delay). But bulk of the logic should remain the same so most if it is ready for review, don't expect much change

File: src/backend/blue-green-deploy.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    const [hostName, port] = host.split(":");
+    const newPort = Number(port) === basePort ? basePort + 1 : basePort;
+    const newImageName = baseImageName + '-' + newVersion;
+    await execSSH(hostName, `cd Expense-Tracker-v2 && git pull && cd ${dockerPath} && docker build -t ${newImageName} . && docker run -dp 127.0.0.1:${newPort}:${basePort} ${newImageName}`);
+    return `${hostName}:${newPort}`;
+}
+
+const provisionGreenDeployments = async (nodes, baseImageName, basePort, newVersion, dockerPath) => {
+    const newNodes = (await Promise.all(nodes.map(async (it) => await startNewInstance(it, baseImageName, basePort, newVersion, dockerPath)))).map((it) => ({ dial: it }));
+    await fetchWithErrorHandling("http://127.0.0.1:2019/config/apps/http/servers/srv0/routes/0/handle/0/routes/0/handle/0/", { upstreams: newNodes });
```
Comment (thebowenfeng) [inline src/backend/blue-green-deploy.js:50]: Caddy [does not reload](https://caddy.community/t/unable-to-reload-in-no-time-like-nginx-does/15034/2) until all existing requests are processed. It is safe to immediately remove all old nodes from LB.

**Comment by cding21** [comment]
I like this

**Comment by cding21** [comment]
Closing due to stale, may reopen in future.

**Pull request closed without merging**